crates/zkvm-jetpack/src/jets/mega_jets.rs

```diff
@@ -0,0 +1,165 @@
+use nockvm::interpreter::Context;
+use nockvm::jets::util::slot;
+use nockvm::jets::Result;
+use nockvm::noun::{IndirectAtom, Noun, D};
+
+use crate::form::math::bpoly::*;
+use crate::form::poly::*;
+use crate::hand::handle::*;
+use crate::hand::structs::{HoonMap, HoonMapIter};
+use crate::jets::utils::jet_err;
+use crate::noun::noun_ext::NounExt;
+
+use crate::form::mega::{brek, MegaTyp};
+use crate::form::math::base::bpow;
+
+
+fn zero_bpoly() -> BPolyVec {
+    BPolyVec::from(vec![0u64])
+}
+
+fn ones_bpoly(len: usize) -> BPolyVec {
+    BPolyVec::from(vec![1u64; len])
+}
+
+pub fn mp_substitute_mega_jet(context: &mut Context, subject: Noun) -> Result {
+    let sam = slot(subject, 6)?;
+    let stack = &mut context.stack;
+
+    let [p_noun, trace_evals_noun, height_noun, chal_map_noun, dyns_noun, com_map_noun] =
+        sam.uncell()?;
+
+    let Ok(trace_evals) = BPolySlice::try_from(trace_evals_noun) else {
+        return jet_err::<Noun>();
+    };
+    let Ok(height_atom) = height_noun.as_atom() else {
+        return jet_err::<Noun>();
+    };
+    let Ok(height) = height_atom.as_u64() else {
+        return jet_err::<Noun>();
+    };
+    let height_usize = height as usize;
+
+    let Ok(dyns) = BPolySlice::try_from(dyns_noun) else {
+        return jet_err::<Noun>();
+    };
+
+    let Ok(chal_map) = HoonMap::try_from(chal_map_noun) else {
+        return jet_err::<Noun>();
+    };
+    let Ok(com_map) = HoonMap::try_from(com_map_noun) else {
```

Comment:
Some maps (at least this `com-map`) may be empty, and `HoonMap` fails if `~` is passed. So you should do `HoonMap::try_from(com_map).ok()`, and then use `com_map.and_then(|m| m.get())`